### ISSUE TITLE
PostgreSQL json/jsonb data type error

### ISSUE THREAD

**Issue by francois** [RailsEventStore/rails_event_store#733]
Using the following reproduction script:

```sh
rails new -m https://railseventstore.org/new repro --database=postgresql
cd repro
sed -i '' 's/binary/jsonb/' db/migrate/*.rb
bin/rails db:create db:migrate
cat <<EOF > repro.rb
OrderPlaced = Class.new(RailsEventStore::Event)

stream_name = "order_1"
event = OrderPlaced.new(data: {
  order_id: 1,
  order_data: "sample",
  festival_id: "b2d506fd-409d-4ec7-b02f-c6d2295c7edd"
})

#publishing an event for a specific stream
event_store = RailsEventStore::Client.new(
  mapper: RubyEventStore::Mappers::Default.new(
    serializer: JSON
  )
)
event_store.publish(event, stream_name: stream_name)
EOF

psql --dbname postgres://localhost/repro_development --command "\d event_store_events"
psql --dbname postgres://localhost/repro_development --command "select * from event_store_events"
```

The last two commands output this:

```
                          Table "public.event_store_events"
   Column   |            Type             | Collation | Nullable |      Default
------------+-----------------------------+-----------+----------+-------------------
 id         | uuid                        |           | not null | gen_random_uuid()
 event_type | character varying           |           | not null |
 metadata   | jsonb                       |           |          |
 data       | jsonb                       |           | not null |
 created_at | timestamp without time zone |           | not null |
Indexes:
    "event_store_events_pkey" PRIMARY KEY, btree (id)
    "index_event_store_events_on_created_at" btree (created_at)
    "index_event_store_events_on_event_type" btree (event_type)

id         | e6b8d8ee-7d20-4bec-bf57-efd030b69510
event_type | OrderPlaced
metadata   | "{\"timestamp\":\"2020-05-09 16:21:25 UTC\"}"
data       | "{\"order_id\":1,\"order_data\":\"sample\",\"festival_id\":\"b2d506fd-409d-4ec7-b02f-c6d2295c7edd\"}"
created_at | 2020-05-09 16:21:25.254738
```

Notice that the data and metadata columns are JSON strings, not JSON objects. This prevents any use of the PostgreSQL `->` and `->>` operators to query values from the columns. A "true" json object displays differently:

```sql
INSERT INTO event_store_events(event_type, metadata, data, created_at)
VALUES (
  'OrderPlaced',
  '{}',
  '{"order_id":2,"order_data":{}}',
  current_timestamp
);
SELECT *, jsonb_typeof(data) FROM event_store_events;
```

```
-[ RECORD 1 ]-----------------------------------------------------------------------------------------------------
id           | e6b8d8ee-7d20-4bec-bf57-efd030b69510
event_type   | OrderPlaced
metadata     | "{\"timestamp\":\"2020-05-09 16:21:25 UTC\"}"
data         | "{\"order_id\":1,\"order_data\":\"sample\",\"festival_id\":\"b2d506fd-409d-4ec7-b02f-c6d2295c7edd\"}"
created_at   | 2020-05-09 16:21:25.254738
jsonb_typeof | string
-[ RECORD 2 ]-----------------------------------------------------------------------------------------------------
id           | fbd2a6ef-43a5-4bdf-aeec-f3cc0fcf155a
event_type   | OrderPlaced
metadata     | {}
data         | {"order_id": 2, "order_data": {}}
created_at   | 2020-05-09 12:27:24.424966
jsonb_typeof | object
```

Notice the `jsonb_typeof` column: string and object. Somewhere in the layers of abstraction, this happens:

```ruby
JSON.dump(JSON.dump(order_id: 1))
#=> "\"{\\\"order_id\\\":1}\""
```

This is the root cause of the issue.

If I replace the JSON mapper by a null one, the issue does not manifest:

```ruby
class Null
  def self.dump(obj)
    obj
  end

  def self.load(obj)
    obj
  end
end

event_store = RailsEventStore::Client.new(
  mapper: RubyEventStore::Mappers::Default.new(
    serializer: Null
  )
)
event_store.publish(event, stream_name: stream_name)
```

```
-[ RECORD 1 ]-----------------------------------------------------------------------------------------------------
id         | e6b8d8ee-7d20-4bec-bf57-efd030b69510
event_type | OrderPlaced
metadata   | "{\"timestamp\":\"2020-05-09 16:21:25 UTC\"}"
data       | "{\"order_id\":1,\"order_data\":\"sample\",\"festival_id\":\"b2d506fd-409d-4ec7-b02f-c6d2295c7edd\"}"
created_at | 2020-05-09 16:21:25.254738
-[ RECORD 2 ]-----------------------------------------------------------------------------------------------------
id         | fbd2a6ef-43a5-4bdf-aeec-f3cc0fcf155a
event_type | OrderPlaced
metadata   | {}
data       | {"order_id": 2, "order_data": {}}
created_at | 2020-05-09 12:27:24.424966
-[ RECORD 3 ]-----------------------------------------------------------------------------------------------------
id         | 8e745720-cbc2-4051-80dc-7493115928a8
event_type | OrderPlaced
metadata   | {"timestamp": "2020-05-09T16:31:18.672Z"}
data       | {"order_id": 1, "order_data": "sample", "festival_id": "b2d506fd-409d-4ec7-b02f-c6d2295c7edd"}
created_at | 2020-05-09 16:31:18.716307
```

I debugged the library, but have not been able to pinpoint the exact place where the 2nd dump happens.

If this issue is fixed, there will be a need to migrate existing data to fix the problem. This SQL code can be used to fix the data:

```sql
BEGIN;
UPDATE event_store_events
SET data = (('[' || data::text || ']')::json->>0)::jsonb
WHERE jsonb_typeof(data) = 'string';
UPDATE event_store_events
SET metadata = (('[' || metadata::text || ']')::json->>0)::jsonb
WHERE jsonb_typeof(metadata) = 'string';

SELECT * FROM event_store_events;
```

This corrects the data so that everything is an object:

```
-[ RECORD 1 ]----------------------------------------------------------------------------------------------
id         | fbd2a6ef-43a5-4bdf-aeec-f3cc0fcf155a
event_type | OrderPlaced
metadata   | {}
data       | {"order_id": 2, "order_data": {}}
created_at | 2020-05-09 12:27:24.424966
-[ RECORD 2 ]----------------------------------------------------------------------------------------------
id         | 8e745720-cbc2-4051-80dc-7493115928a8
event_type | OrderPlaced
metadata   | {"timestamp": "2020-05-09T16:31:18.672Z"}
data       | {"order_id": 1, "order_data": "sample", "festival_id": "b2d506fd-409d-4ec7-b02f-c6d2295c7edd"}
created_at | 2020-05-09 16:31:18.716307
-[ RECORD 3 ]----------------------------------------------------------------------------------------------
id         | e6b8d8ee-7d20-4bec-bf57-efd030b69510
event_type | OrderPlaced
metadata   | {"timestamp": "2020-05-09 16:21:25 UTC"}
data       | {"order_id": 1, "order_data": "sample", "festival_id": "b2d506fd-409d-4ec7-b02f-c6d2295c7edd"}
created_at | 2020-05-09 16:21:25.254738
```

The UPDATE statements above are idempotent: running them again against the same data will not cause problems.

Hope that helps!

**Comment by greveritt** [comment]
What version of the gem are you using, @francois? This sounds like a duplicate of #651, unless I'm missing something.

**Comment by francois** [comment]
@greveritt, I use 1.0:

```ruby
gem "rails_event_store", "~> 1.0"
```

You are correct: this is a duplicate of #651. I'll let you choose which issue should be kept and which should be removed.

**Comment by mpraglowski** [comment]
JSON mapper that handled this issue has been implemented in PR #680
